File: kvdom/kvdom-dom/src/wasm32Main/kotlin/lt/petuska/kvdom/dom/Event.kt
package lt.petuska.kvdom.dom

import kotlinx.wasm.jsinterop.Arena
import kotlinx.wasm.jsinterop.JsValue
import kotlinx.wasm.jsinterop.Object

actual open class Event(arena: Arena, index: Object) : JsValue(arena, index) {
  actual fun preventDefault() {
    js_preventDefault(arena, index)
  }
  
  actual fun initEvent(type: String, bubbles: Boolean, cancelable: Boolean) {
    TODO()
  }
  
  actual fun stopPropagation() {
    TODO()
  }
  
}

@SymbolName("kvdom_Event_preventDefault")
private external fun js_preventDefault(
  arena: Arena,
  index: Object,
)

File: kvdom/kvdom-dom/src/wasm32Main/resources/Event.js
konan.libraries.push({
    kvdom_Event_preventDefault: function (arena, obj) {
        kotlinObject(arena, obj).preventDefault();
    },
    kvdom_Event_stopPropagation: function (arena, obj) {
        kotlinObject(arena, obj).stopPropagation();
    },
    kvdom_Event_initEvent: function (arena, obj, typePtr, typeLen, bubbles, cancelable) {
        kotlinObject(arena, obj).initEvent(toUTF16String(typePtr, typeLen, bubbles === 1, cancelable === 1));
    },
});
